SnowplowTracker/SnowplowTracker/Storage/EventStore.cs

```diff
@@ -0,0 +1,296 @@
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="SnowplowTracker.Storage.EventStore"/> class.
+		/// </summary>
+		public EventStore() {
+			Open ();
+			payloadQueueConsumer = new Thread(new ThreadStart(PayloadQueueConsumer));
+			payloadQueueConsumer.Start ();
+		}
+
+		/// <summary>
+		/// Opens a connection and create's the events tables if required.
+		/// </summary>
+		public void Open() {
+			if (!IsDatabaseOpen ()) {
+				Log.Verbose("EventStore: opening database: " + DATABASE_NAME);
+				dbConnection = new SqliteConnection (DATABASE_NAME);
+				dbConnection.Open();
+
+				IDbCommand command = dbConnection.CreateCommand ();
+
+				// Set database mode to Write Ahead Logging
+				command.CommandText = "PRAGMA JOURNAL_MODE = WAL;";
+				command.ExecuteNonQuery();
+
+				// Set database mode to Serialized for safe multi-threading
+				command.CommandText = "PRAGMA SQLITE_THREADSAFE = 1;";
+				command.ExecuteNonQuery();
+
+				// Create the Table if it does not exist
+				command.CommandText = CREATE_TABLE;
+				command.ExecuteNonQuery();
+
+				Log.Verbose(" + Database open status: " + IsDatabaseOpen ());
+
+				// Close command
+				command.Dispose();
+				command = null;
+			}
+		}
+
+		/// <summary>
+		/// Closes the database connection if is currently active.
+		/// </summary>
+		public void Close() {
+			if (IsDatabaseOpen ()) {
+				dbConnection.Close();
+				dbConnection.Dispose();
+				dbConnection = null;
+			}
+
+			// Close consumers
+			payloadQueueConsumerContinue = false;
+		}
+
+		// --- Database Functions
+
+		/// <summary>
+		/// Adds the event to a concurrent queue.
+		/// </summary>
+		/// <param name="payload">Payload.</param>
+		public void AddEvent(TrackerPayload payload) {
+			payloadQueue.Enqueue (payload);
+		}
+
+		/// <summary>
+		/// Adds an event payload
+		/// </summary>
+		/// <returns><c>true</c>, if event was added, <c>false</c> otherwise.</returns>
+		/// <param name="payload">An event payload</param>
+		private void PayloadQueueConsumer() {
+			TrackerPayload payload = payloadQueue.Dequeue ();
+			bool result = false;
+			if (IsDatabaseOpen ()) {
+				byte[] bytes = Utils.SerializeDictionary (payload.GetDictionary());
+				if (bytes != null) {
+					// Build Insert Command
+					IDbCommand command = dbConnection.CreateCommand ();
+					command.CommandText = INSERT_EVENT + "(@payload)";
+					SqliteParameter param = new SqliteParameter ("@payload", DbType.Binary);
+					param.Value = bytes;
+					command.Parameters.Add (param);
+					
+					// Check whether insert succeeded
+					result = command.ExecuteNonQuery () > 0;
+					
+					// Close and dispose of resources
+					command.Dispose();
+					command = null;
+				}
+			}
+			Log.Verbose("EventStore: Event add result: " + result);
+
+			// Loop forever or until we decide to stop
+			if (payloadQueueConsumerContinue) {
+				PayloadQueueConsumer ();
```

Comment:
Ahh really! I will make sure to fix the emitter loop as well.  I could also use a timer with a 0ms delay to reschedule it as an informal tail recursion.